File: 2022/day16/day16.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <regex>
#include <map>

struct Valve
{
    Valve(int flowIn, std::vector<std::string> connectedValvesIn) : flow(flowIn), connectedValves(connectedValvesIn) {}
    Valve() {}

    int flow;
    std::vector<std::string> connectedValves;
};

void part1(std::map<std::string, Valve> valves) {
    std::vector<std::map<std::string, int>> dp(31);

    dp[0]["AA"] = 0;

    for(int i = 1; i <= 30; i++) {
        for(auto &pair: dp[])
    }

    int max = 0;
    for(auto &pair: dp[30]) {
        max = std::max(max, std::get<1>(pair));
    }
    std::cout << "Part 1: " << max << std::endl; 
}

int main()
{
    std::ifstream infile("input.txt");
    std::string tmp;
    std::regex pattern("Valve (..) has flow rate=(..?); tunnels? leads? to valves? (.*)");

    std::map<std::string, Valve> valves;

    while (std::getline(infile, tmp))
    {
        std::smatch matches;
        std::regex_search(tmp, matches, pattern);

        std::string valveCode = matches[1];

        int flowRate = std::stoi(matches[2]);

        std::stringstream ss(matches[3]);
        std::vector<std::string> connectedValves;
        while (std::getline(ss, tmp, ','))
        {
            connectedValves.push_back(tmp);
        }

        valves[valveCode] = Valve(flowRate, connectedValves);
    }

    part1(valves);

}

File: 2022/day16/day16.py
from pathlib import Path
import regex as re

p = Path(__file__).with_name("input.txt")
file = open(p)

flowRates = {}
tunnels = {}
dp = [{}] * 32

for line in file.readlines():
    matches = re.findall("Valve (\w\w).*rate=(\d+); tunnels? leads? to valves? (.*)", line)

    valve = matches[0][0]

    flowRate = int(matches[0][1])
    flowRates[valve] = flowRate

    tunnelsTo = matches[0][2].split(", ")
    tunnels[valve] = tunnelsTo

start = 'AA'
dp[0]['AA'] = (0, 0, []) # flowRate, total, valvesOpen

for i in range(1, 31):
    valves = dp[i-1]
    valvesForCurrentLevel = {}

    for (valve, maxAtPreviousLevel) in valves.items():
        (currentFlowRate, currentTotal, valvesOpen) = maxAtPreviousLevel

        newTotal = currentTotal + currentFlowRate
        
        # Two Decisions: we open valve, or we move somewhere else
        if True:
            targetValve = valve
            if not valve in valvesOpen:
                newFlowRate = currentFlowRate + flowRates[targetValve]
                newValvesOpen = valvesOpen + [targetValve]
            else:
                newFlowRate = currentFlowRate
                newValvesOpen = valvesOpen

            newItem = (newFlowRate, newTotal, newValvesOpen)

            if not targetValve in valvesForCurrentLevel:
                valvesForCurrentLevel[targetValve] = newItem
            elif valvesForCurrentLevel[targetValve] == newFlowRate and valvesForCurrentLevel[targetValve][1] < newTotal:
                valvesForCurrentLevel[targetValve] = newItem
            elif valvesForCurrentLevel[targetValve][0] < newFlowRate:
                valvesForCurrentLevel[targetValve] = newItem


        for targetValve in tunnels[valve]:
            newItem = (currentFlowRate, newTotal, valvesOpen)

            if not targetValve in valvesForCurrentLevel:
                valvesForCurrentLevel[targetValve] = newItem
    
            elif valvesForCurrentLevel[targetValve] == newFlowRate and valvesForCurrentLevel[targetValve][1] < newTotal:
                valvesForCurrentLevel[targetValve] = newItem
            elif valvesForCurrentLevel[targetValve][0] < newFlowRate:
                valvesForCurrentLevel[targetValve] = newItem

                

    dp[i] = valvesForCurrentLevel
    print("Minute", i)
    print(dict(sorted(dp[i].items())))

part1 = max(dp[30].values(), key=lambda p: p[1])[1]
print("Part 1", part1)